infrastructure/code/code.go

```diff
@@ -32,13 +32,34 @@ type ScanMetrics struct {
 	lastScanFileCount         int
 }
 
+type ScanStatus struct {
+	// finished channel is closed once the scan has finished
+	finished chan struct{}
+
+	// isRunning is true when the scan is either running or waiting to run, and changed to false when it's done
+	isRunning bool
+
+	// isPending is true when the scan is currently waiting for a previous scan to finish
+	isPending bool
+}
+
+func NewScanStatus() *ScanStatus {
+	return &ScanStatus{
+		finished:  make(chan struct{}),
+		isRunning: false,
+		isPending: false,
+	}
+}
+
 type Scanner struct {
-	BundleUploader *BundleUploader
-	SnykApiClient  snyk_api.SnykApiClient
-	errorReporter  error_reporting.ErrorReporter
-	analytics      ux2.Analytics
-	ignorePatterns []string
-	mutex          sync.Mutex
+	BundleUploader  *BundleUploader
+	SnykApiClient   snyk_api.SnykApiClient
+	errorReporter   error_reporting.ErrorReporter
+	analytics       ux2.Analytics
+	ignorePatterns  []string
+	mutex           sync.Mutex
+	scanStatusMutex sync.Mutex
```

Comment:
Is it really necessary to have a second mutex here?

Reply:
The 1st mutex is locking a section inside `files` on a section that updates the `ignorePatterns` field, the 2nd one is for synchronizing new scan requests